packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts

```diff
@@ -0,0 +1,418 @@
+
+  const mixedRouteIsAllV3 = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+    return r.pools.every(SmartRouter.isV3Pool)
+  }
+  const mixedRouteIsAllV2 = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+    return r.pools.every(SmartRouter.isV2Pool)
+  }
+  const mixedRouteIsAllStable = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+    return r.pools.every(SmartRouter.isStablePool)
+  }
+
+  // Check if a mixed route is actually pure v2, v3 or stable route.
+  // If that's the case then let the corresponding command adder to handle the encode
+  if (mixedRouteIsAllV3(route)) {
+    addV3Swap(
+      planner,
+      {
+        ...trade,
+        routes: [
+          {
+            ...route,
+            type: RouteType.V3,
+          },
+        ],
+        inputAmount,
+        outputAmount,
+      },
+      options,
+      routerMustCustody,
+      payerIsUser,
+    )
+    return
+  }
+  if (mixedRouteIsAllV2(route)) {
+    addV2Swap(
+      planner,
+      {
+        ...trade,
+        routes: [
+          {
+            ...route,
+            type: RouteType.V2,
+          },
+        ],
+        inputAmount,
+        outputAmount,
+      },
+      options,
+      routerMustCustody,
+      payerIsUser,
+    )
+    return
+  }
+  if (mixedRouteIsAllStable(route)) {
+    addStableSwap(
+      planner,
+      {
+        ...trade,
+        routes: [
+          {
+            ...route,
+            type: RouteType.STABLE,
+          },
+        ],
+        inputAmount,
+        outputAmount,
+      },
+      options,
+      routerMustCustody,
+      payerIsUser,
+    )
+    return
+  }
+
+  invariant(trade.tradeType === TradeType.EXACT_INPUT, 'Exact output is not supported for mixed route trade')
+
+  const sections = SmartRouter.partitionMixedRouteByProtocol(route)
+
+  let outputToken: Currency | undefined
+  let inputToken = inputAmount.currency.wrapped
+
+  for (let i = 0; i < sections.length; i++) {
+    const section = sections[i]
+    const nextSection = sections[i + 1] ?? []
+    const isFirstSection = i === 0
+    const isLastSection = i === sections.length - 1
+
+    const nextIsV2 = nextSection.length && nextSection.every(SmartRouter.isV2Pool)
+
+    const getRecipient = (): Address => {
+      if (isLastSection) return recipient
+      if (nextIsV2) {
+        const address = getPoolAddress(nextSection[0])
+        if (!address) throw new Error('unknown v2 pool address')
+        return address
+      }
+      return ROUTER_AS_RECIPIENT
+    }
+    const currentRecipient = getRecipient()
+
+    // Now, we get output of this section
+    outputToken = SmartRouter.getOutputOfPools(section, inputToken)
+    const newRoute = SmartRouter.buildBaseRoute([...section], inputToken, outputToken)
+    inputToken = outputToken.wrapped
+
+    const payByUser = payerIsUser && isFirstSection
+
+    const inAmount = isFirstSection ? amountIn : CONTRACT_BALANCE
+    const outAmount = isLastSection ? amountOut : 0n
```

Comment:
even not last section should also calculate minOutAmout

Reply:
nah we only do the estimation so we don't know how much token we will get exactly for each hop. That's why for the intermediate swap we don't specify the amount out

Reply:
yes, the test have updated to the input/output amount now